File: typescript/src/majorityElement/majorityElement.ts
const majorityElement = (nums: number[]): number => {
    let votes = 0;
    let candidate = 0;
    for(let num of nums){
        if(votes == 0){
            candidate = num;
        }
        if(num == candidate){
            votes++;
        } else {
            votes--;
        }
    }
    return candidate;
};

const nums = [2,2,1,1,1,2,2];
console.log(majorityElement(nums));

File: python/majority_element/majority_element.py
from typing import List


class Solution:
    # Moore voting algorithm
    def majorityElement(self, nums: List[int]) -> int:
        votes = 0
        candidate = 0
        for num in nums:
            if votes == 0:
                candidate = num
            if num == candidate:
                votes += 1
            else:
                votes -= 1
        return candidate


nums = [2, 2, 1, 1, 1, 2, 2]
print(Solution().majorityElement(nums))
